### ISSUE TITLE
can this be safely tested without risk of bsod?

### ISSUE THREAD

**Issue by mailinglists35** [lundman/ZFSin#20]

**Comment by gamlerhart** [comment]
No...it's currently at proof of concept stage. 
You risk BSODs and other issues. Don't put it on any production machine =)

**Comment by lundman** [comment]
It is trivial to test it on a VM, which is why BSOD won't matter in the least, since you just reboot the VM if that happens.   Now it's been *hours* and *hours* since I had a BSOD myself, but yeah, if you put it on a real machine, you will have to deal with it if it crashes, and how to remove the ZFSin.sys file from recovery boot.

**Comment by mailinglists35** [comment]
is there any documented procedure how to report bsods if I were starting to test it? full memory dumps or just minidump.dmp uploads? need to install windows debugging tools?

**Comment by mailinglists35** [comment]
ideall i'd like to experiment it on physical hardware not on vm

**Comment by lundman** [comment]
Probably more of a `I ran command X then Y and it crashed` - and if I can repeat that in the debugger with VM, I can see what goes wrong. Never personally found/figured out Windows crash reports (assuming it has them), but one day I'll learn :)

**Comment by mailinglists35** [comment]
>  Never personally found/figured out Windows crash reports (assuming it has them), but one day I'll learn :)

it has. keywords ["WinDbg" and "!analyze -v"](https://blogs.technet.microsoft.com/markrussinovich/2010/12/12/the-cases-of-the-blue-screens-finding-clues-in-a-crash-dump-and-on-the-web/)

when you have time, i recommend [windows internals by mark russinovich](https://docs.microsoft.com/en-us/sysinternals/learn/windows-internals). this guy is God. 
https://www.microsoftpressstore.com/store/windows-internals-part-1-system-architecture-processes-9780735684188 (7th edition part 1) you likely will need part 2 when it will be released (7th edition covers windows 10, 6th edition windows 7)
````
This book will help you:

·        Understand the Window system architecture and its most important entities, such as processes and threads

·        Examine how processes manage resources and threads scheduled for execution inside processes

·        Observe how Windows manages virtual and physical memory

·        Dig into the Windows I/O system and see how device drivers work and integrate with the rest of the system

·        Go inside the Windows security model to see how it manages access, auditing, and authorization, and learn about the new mechanisms in Windows 10 and Server 2016
````


https://www.microsoftpressstore.com/store/windows-internals-part-2-9780735665873 (6th edition part 2)
````
In Part 2, you’ll examine:

Core subsystems for I/O, storage, memory management, cache manager, and file systems
Startup and shutdown processes
Crash-dump analysis, including troubleshooting tools and techniques
````

if you want the book(s), i'd be happy to make my first donation back for all the free opensource stuff i've used over the years without paying a cent.

**Comment by lundman** [comment]
> it has. keywords "WinDbg" and "!analyze -v"

Ah I use that every time now, since I launch the remote machine with the debugger. Didn't know you could use that on a machine that didn't have remote-debugging installed.

I will check out the book recommended and do a bit more learning about Windows. I was sort of hoping that after I had done the "easy Unix part", someone from Windows could help with the "hard Windows part", since the "easy" and "hard" words would presumably be swapped for them. :)

**Comment by lundman** [comment]
While reading this book, we should check into:

* Windows can start threads deferred - look at removing `taskq` mutex fix for OsX inability to do so.
* LARGE_PAGE allocs of 2MB, possibly kmem should use that quantum instead of pagesize

**Comment by mailinglists35** [comment]
>  Didn't know you could use that on a machine that didn't have remote-debugging installed.

I don't think that works on the console of the same running machine, but it works for sure for posts-mortem analyzing of crash dumps (either minidumps or full memory dumps)

**Comment by mailinglists35** [comment]
> someone from Windows could help with the "hard Windows part"

the only free public work I've found about native windows filesystem integration are two projects - try contacting the author of [ext2ifs](http://www.fs-driver.org/author.html) and look at the more up to date [ext2fsd project](http://www.ext2fsd.com/) ; perhaps [reactos](http://new.reactos.org/participation#devanchor) guys could know more intimate stuff about windows

I'm seeing some people commenting on a [stackexchange discussion](https://stackoverflow.com/questions/6789392/windows-file-systems), at least one of them is selling a related windows storage driver product, maybe he could give some hints ([A different approach](http://www.xtreemfs.org/download_windows.php) has been taken by the XtreemFS project by using the cbfs framework (non-free), one of the guys commenting there appears to own the company behind cbfs framework)

The author of [this site](http://www.acc.umu.se/~bosse/) seems to have some knowledge of developing stuff for windows storage. Try pinging him.

[These guys](http://openafs.org/windows.html) also wrote native windows filesystem driver, as well as [this guy](http://www.chrysocome.net/contact). 

btw, is zfsonwin building as an [installable file system minifilter driver](https://docs.microsoft.com/en-us/windows-hardware/drivers/ifs/filter-manager-and-minifilter-driver-architecture), right?

and finally Microsoft has some [sample code on github](https://github.com/Microsoft/Windows-driver-samples/tree/master/filesys)

**Comment by fire** [comment]
I've been testing ZFSin on Windows for a few weeks without crashing Windows. 

Try the latest.

**Comment by vmlemon** [comment]
Not 100% perfect, but a picture says a thousand words:
![image](https://user-images.githubusercontent.com/35642/54095351-cfb2b500-439e-11e9-8383-77ec29939444.png)

Here's Windows 8 RTM, running under VirtualBox (which is installed on an NTFS VHD, on a deduplicated, compressed ZFS dataset, stored on a 2TB USB HDD, shared with a FreeBSD installation), on a crappy laptop (only 4GB of internal RAM, 32GB internal eMMC, Celeron CPU), running a not-quite-up-to-date version of Windows 10, and the latest 0.17 release of OpenZFS.

The VM's disk image is stored in the root of the ZFS dataset, as a dynamically-expanding VHD, as is the ISO that was used to install Windows 8. 

The whole process was quick, and there wasn't a single hang, or BSOD... 

I did have another comment prepared, regarding various caveats, but I'll probably post that, later. :)

**Comment by vmlemon** [comment]
(With earlier versions, that feat would have failed miserably, with errors about the SATA controller being unreliable, and I probably would have received a few BSODs, just trying to download the Windows ISO, onto said ZFS dataset)

**Comment by vmlemon** [comment]
You can still get BSODs, and stalls/crashes, under certain circumstances, with various hardware/software configurations (e.g. when using ZPools hosted in NTFS VHDs, on some Seagate external HDDs, with "eco-mode", that insist on spinning themselves down, incessantly, so everything hangs, when doing heavy I/O; or, when trying to unmount a pool, after lots of writing; and, downloading files using Chrome used to cause random BSODs, for some reason), but it's surprisingly-reliable, right now.
